design_documents/contract_verification.md

```diff
@@ -0,0 +1,118 @@
+# Cairo Contract Verification via Block Explorers
+
+## Context
+
+Cairo smart contracts deployed to Starknet are only visible as a Cairo Bytecode and their's ABIs, which is difficult to decipher and understand. To increase readability of a smart contract, it's source code can also be made public, however in order to ensure trustlessness, the source code vs resulting Cairo Bytecode verification can be performed. Such verification and its results can be performed and made public by popular Blockchain Explorers.
+
+## Goal
+
+This proposal includes an extension to `sncast` utility enabling a contract owner to perform contract verification against a selected Blockchain Explorer API. We propose to create a first, reference implementation for the Voyager APIs.
+
+## Proposed Solution
+
+We propose to design a dedicated `verify` command for the `sncast` tool, and add a mechanism whereby this command can be implemented for various Blockchain Explorers.
+
+### `sncast` utility command - `verify`
+
+Command name: `verify`
+
+The `verify` command will perform following actions:
+- select the verifiation logic to use (based on `--verifier` parameter)
+- pick the selected scarb workspace source code from local filesystem (based on `Scarb.toml` file)
+- for the selected verifier it will upload the Scarb workspace contracts source code to the verifier's API
+- call the verifier's API to trigger verification of the uploaded source code
+- wait for results of the verification API call
+- respond to the user with the results 
+
+#### Parameters
+
+#### `--path-to-scarb-toml, -s <PATH>`
+
+Optional.
+Path to `Scarb.toml` file.
+If supplied, cast will not look for `Scarb.toml`` file in current (or parent) directory, but will use this path instead.
+
+#### `--contract-name`
+
+Required.
+Name of the contract to be submitted for verification.
+
+#### `--contract-address`
+
+Required.
+Address of the contract to be submitted for verification.
+
+#### `--verifier <VERIFIER NAME>`
+
+Optional.
+Specifies the Blockchan Explorer to verify with.  
+Default: `voyager`
+
+Options are: 
+ - voyager
+ - starkscan
+
+#### `--verifier-url <URL>`
+
+Optional.
+Specifies the Blockchain Explorer's Verification API URL ().  
+Default: the default API URL for selected verifier (eg. https://goerli.voyager.online/) (???)
+
+Options are: 
+ - `https://voyager.online/` - for verification at mainnet
+ - `https://goerli.voyager.online/` -  for verification on testnet
+
+### Voyager API plugin
+
+A sample request to the voyager API end point will look as follows: 
+```rust
```

Comment:
I'm curious if that would be possible, but that would indeed be cool

Reply:
It would be better on our end, maintaining would be easier that way.

Reply:
will work on that what type of responses the sncast will `expect`, will update the same in this PR

Reply:
@cptartur @THenry14 we are assuming we have no influence on the APIs provided by any of the block explorers - ie. Voyager verification APIs are there already and are different form StarkScan APIs, etc. Hence - the idea of "adapters" in `sncast`, that encapsulate the nuances of each specific API and expose a generic interface (trait?) to the `verify` command implementation. And for this iteration if the feature, we would be implementing Voyager adapter only. Does that make sense?

Reply:
Perhaps enforcing some api standard is a bit overkill and we can go with "adapters" approach. Preferably we'd still have some minimal interface that explorers should implement if they want to integrate with `sncast`. Maybe it can be just at a trait level and explorers should be allowed to send whatever request format they want as long as they can then return it as whatever rust structure we expect. 

My concern is just if we have multiple explorers supported I'd like the outputs presented to the user in the CLI to be exactly the same, no matter which one is used.

I'm understand that initially the response will basically be `verified/not verified` but I'm assuming this might get expanded in the future?

Reply:
Added `### Contract Verification interface` section to propose a generic request/response format. I hope this makes the idea more clear?

Reply:
It does, thanks. Resolving this discussion, will add comments to the updated section if I have any.